### ISSUE TITLE
[Feature Request] Support .nvmrc for `pnpm env`

### ISSUE THREAD

**Issue by kylemh** [pnpm/pnpm#4471]
Just learned about `pnpm env` from [this Twitter thread](https://twitter.com/jaredpalmer/status/1506634640509083649)

- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

**Describe the user story**

If I have an `.nvmrc` file, I'd like to be able to run `pnpm env` and have that Node version automatically install (if needed) and be switched to.

**Comment by xeho91** [comment]
If this feature request would pass, how about using the `engines` key from `package.json`, where we can specify node version, instead of another file `.nvmrc`?

```json
{
  "engines": {
    "node": "16.14.2"
  }
}
```

**Comment by zkochan** [comment]
No, we cannot use the engines field because if you publish such package and you will install it with `16.14.3`, you'll get a warning. Or even a crash.

**Comment by tsndr** [comment]
Support for `.nvmrc`files is the only think keeping be from switching to pnpm.

**Comment by privatenumber** [comment]
Alternatively, a `node-version-file` rc option like [`@actions/setup-node`](https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#node-version-file).

**Comment by zkochan** [comment]
Why should pnpm read from `.nvmrc`? pnpm reads settings from `.npmrc`.

You can specify the node version via [use-node-version](https://pnpm.io/npmrc#use-node-version) in `.npmrc`.

**Comment by privatenumber** [comment]
It doesn't necessarily have to be `.nvmrc`, but there seems to be a convention to have a file containing just the Node.js version amongst popular Node.js managers (eg. nvm, fnm, n, `@actions/setup-node`, etc).

If it's in a file of its own, it can either be `cat`'d into the "install Node" command or symlinked to the appropriate config file path used by other Node.js managers.

**Comment by adam-zielonka** [comment]
Hello, you're missing also `.node-version` file that it use for some node version managers. 

It's so stupid that there is no one way to do it.

**Comment by weyert** [comment]
and `.tool_versions` for `asdf`

**Comment by privatenumber** [comment]
I think it would be fine if pnpm could just support one file format.

Every other file name can be supported via symlink.

**Comment by adam-zielonka** [comment]
I wrote the message before try this node version management feature. I change my mind. It is not necessary to support other ways for determining node version, because it is invisible for developers. I just added node version in pnpm way in one project and I am sure that most of my colleagues will not realise about change, when they start working on this project. And CI it's just simpler, one image with pnpm to maintenance.

**Comment by privatenumber** [comment]
If you're using GitHub Actions, here's a relevant CI limitation with pnpm setting up Node.js: https://github.com/pnpm/action-setup/issues/70

**Comment by shadowspawn** [comment]
For interest, node version managers and products supporting `.node-version`: https://github.com/shadowspawn/node-version-usage

**Comment by Kytech** [comment]
NVM is quite popular, so supporting `.nvmrc` is probably worthwhile. I think `.nvmrc` and `.node-version` are the two most popular files used for specifying a node version.

> And would it work with pnpm env use without args? Or just pnpm env?

I'd think it would make most sense to have the behavior of reading from a node version file be for a no-arg use of `pnpm env`. (Think nvm's usage pattern). Specifying argument(s) should be for cases where you'd want to override the behavior or node version that the project specifies in its file. I know its very common for users of node version managers to use a shell integration that will auto-switch the node version if they enter a folder with a node version file. This is also another point in favor of using a separate node version file instead of just in the .npmrc file - this would make it much easier for people to adopt pnpm for an auto-switch shell integration script.

**Comment by sschuberth** [comment]
> how about using the `engines` key from `package.json`, where we can specify node version

That would be https://github.com/pnpm/pnpm/issues/7211, BTW.

**Comment by affanshahid** [comment]
I think `.node-version` and `.nvmrc` support is crucial, otherwise it is difficult to use pnpm for managing node versions in a project that does not use it for dep management without polluting it with an extra file. Most projects will already have either of these checked in. So until pnpm supports them I need to either add another file, change the global version, or use a separate version manager. The last option defeats the purpose of using pnpm as the version manager.

**Comment by jamesknelson** [comment]
For anybody looking for a way to automatically install/use the node version that pnpm *does* read (in `.npmrc`), assuming you've only got a single line in `.npmrc` and it is just to specify the node version, you can use this bash one liner:

```
pnpm env use --global $(grep "use-node-version=" .npmrc | cut -d'=' -f2)
```

**Comment by sdavids** [comment]
You might want to support `devEngines.node` instead of `engines.node`:


https://docs.npmjs.com/cli/v11/configuring-npm/package-json#devengines:

---


There are lots of projects out there already having a `.nvmrc` and/or `.node-version` file though—it might make sense to also support those files as well.

Beware that `.nvmrc` supports comments though: https://github.com/nvm-sh/nvm/commit/29dce5edfd0976f9a1728c5746715c24061fd404

**Comment by zkochan** [comment]
Updated the implementation in pnpm/pnpm#15247 after reviewing the current pnpm runtime flow. `.nvmrc` is now an additional project runtime source for pnpm's context-aware global `node` shim. It no longer changes `pnpm env use`.

The shim searches upward for the nearest runtime declaration, keeps `devEngines.runtime` and `engines.runtime` authoritative over `.nvmrc` in the same directory, and provisions the selected Node.js version through pnpm's managed runtime store.

---
Written by an agent (Codex, GPT-5).
